doc/manual/contracts.md

```diff
@@ -0,0 +1,635 @@
+  if value == null then
+    value 
+  else
+    contracts.apply contract label value in
+// succeeds
+null | Nullable Num
+// succeeds too
+1 | Nullable Num
+```
+
+## Compound contracts
+
+Contracts can be nicely composed into larger contracts. This is a big upside
+over just using the classic assertions (standard functions and a simple abort
+primitive). In this section, we review the different ways of assembling existing
+contracts together into new ones.
+
+### Records
+
+Record contracts are fundamental in a configuration language like Nickel. They
+encode the data schema of a configuration. A record contract is just a record
+literal with contract annotations and whose field definitions are usually
+missing:
+
+```nickel
+let MyConfig = {
+  path | Str,
+ 
+  connection | {
+    server_port | Port,
+    host | Str,
+  }
+} in
+{
+  path = "/foo/bar",
+  connection = {
+    server_port = 48484,
+    host = "localhost"
+  }
+} | MyConfig
+```
+
+`MyConfig` is a schema describing a configuration with:
+
+- a field `path` that must be a string
+- a sub-record `connection`, which has a `server_port` and a `host` field that
+  are respectively a port number and a string.
+
+See how we've combined contracts: while defining the `MyConfig` record contract,
+we wrote another inline record contract for the `connection` field, and we
+reused the `Port` contract defined earlier. Composition is this ability of using
+existing contracts seamlessly as building blocks for new contracts.
+
+#### Metadata
+
+Additionally to defining the contracts of fields, record contract can also
+attach metadata (see [merging](./merging.md)) to fields, such as a documentation
+or default value:
+
+```
+nickel>let MyConfig = {
+    foo | doc "This documentation will propagate to the final value!"
+        | Str
+        | default = "foo",
+    bar | Num,
+}
+nickel>let config | MyConfig = {bar = 2}
+nickel> builtins.serialize `Json config
+"{
+  "bar": 2,
+  "foo": "foo"
+}"
+nickel>
+nickel>:query config.foo
+* contract: Str
+* default: "foo"
+* documentation: This documentation will propagate to the final value!
+```
+
+#### Open contracts
+
+By default, contracts are closed, meaning that additional fields are forbidden:
+
+```
+nickel>let Contract = {foo | Str}
+nickel>{foo = "a", bar = 1} | Contract
+error: contract broken by a value [extra field `bar`].
+[..]
+```
+
+If you want to allow additional fields, append `, ..` after the last field
+definition to define an open contract:
+
+```
+nickel>let Contract = {foo | Str, ..}
+nickel>{foo = "a", bar = 1} | Contract
+{ foo = <contract,value="a">, bar = 1}
+```
+
+#### Giving values for fields
+
+While most record contracts don't have field definitions, they can. In fact,
+records contracts are not special: any record literal can be used both as a
+contract or a normal value. If a field is defined both in the contract and the
+checked value, the two definitions will be merged in the final result. For
+example:
```

Comment:
I don't recall such a discussion. Initially, inspired from CUE, contracts application was just merging, and there was thus no reason to reject specific records as record contracts. Then it appeared that while retaining some characteristics from merging was nice, record contract application also differed on some aspects, so it is now an asymmetric and slightly specialized version of merge, but it still doesn't impose any restriction on its operands.

Reply:
I'm a little worried that this would lock us in a semantics that we don't know is useful or not. Whereas rejecting it would let us choose later based on experience. It's surprising that

```
{ foo | Int } | { foo = 0 }
```

will pass and and have field `foo = 0`.

The example that you show, however, is reasonable. But the **warning** section also highlights that it's all a little tricky.

I don't know…

Reply:
Initially it appeared to be useful to force a field to a fixed value, like in the example.

Now, I'm not so sure, because we could do that with contracts instead: `{foo | Eq 5 | default = 5}`. It's a bit more verbose, but it removes yet-another-way-of-doing-the-same-thing and fails with the usual blame error rather than a "non mergeable term" error (we could even imagine in the future that it's possible to use literals as contracts directly, as in Racket `{foo | 5 | default = 5}`, or even that a singleton contract automatically set a default value, and just write `{foo | 5}`, but I digress).

Plus the edge cases given in the documentation. So, I tend to think we may be better off by forbidding field definitions altogether. Should we do this right now?